src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,140 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
     @Override
     public void add(T value) {
+        if (head == null) {
+            head = new Node<>(null, value, null);
+            tail = head;
+            size++;
+        } else {
+            Node<T> newNode = new Node<>(tail, value, null);
+            tail.next = newNode;
+            tail = newNode;
+            size++;
+        }
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == size) {
+            add(value);
+            return;
+        }
+        Node<T> getNode = getNode(index);
+        Node<T> getNodePrev = getNode.prev;
+        Node<T> newNode = new Node<>(getNodePrev, value, getNode);
+        if (getNodePrev == null) {
+            head = newNode;
+        } else {
+            getNodePrev.next = newNode;
+        }
+        getNode.prev = newNode;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return getNode(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> nodeToReplace = getNode(index);
+        T lastNodeValue = nodeToReplace.item;
+        nodeToReplace.item = value;
+        return lastNodeValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> nodeToRemove = getNode(index);
```

Comment:
```suggestion
        checkIndex(index);
        Node<T> nodeToRemove = getNode(index);
```
It's better to check infex first and then do other logic